File: src/2025-02-24-deepclaude-rust/javascript/exercises/practice/word-search/word-search.js
//
// This is only a SKELETON file for the 'Word Search' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

class WordSearch {
  constructor(grid) {
    this.grid = grid;
    this.rows = grid.length;
    this.cols = grid[0].length;
  }

  find(words) {
    const directions = [
      [-1, -1], [-1, 0], [-1, 1],  // up-left, up, up-right
      [0, -1],           [0, 1],    // left, right
      [1, -1],  [1, 0],  [1, 1]     // down-left, down, down-right
    ];
    
    const result = {};
    
    words.forEach(word => {
      let found = false;
      
      // Try each starting position
      for (let row = 0; row < this.rows && !found; row++) {
        for (let col = 0; col < this.cols && !found; col++) {
          // Try each direction
          for (const [dx, dy] of directions) {
            if (this.checkWord(word, row, col, dx, dy)) {
              const endRow = row + dx * (word.length - 1);
              const endCol = col + dy * (word.length - 1);
              result[word] = {
                start: [row + 1, col + 1],
                end: [endRow + 1, endCol + 1]
              };
              found = true;
              break;
            }
          }
        }
      }
      
      if (!found) {
        result[word] = undefined;
      }
    });
    
    return result;
  }

  checkWord(word, startRow, startCol, dx, dy) {
    if (
      startRow + dx * (word.length - 1) < 0 ||
      startRow + dx * (word.length - 1) >= this.rows ||
      startCol + dy * (word.length - 1) < 0 ||
      startCol + dy * (word.length - 1) >= this.cols
    ) {
      return false;
    }

    for (let i = 0; i < word.length; i++) {
      const row = startRow + dx * i;
      const col = startCol + dy * i;
      if (this.grid[row][col] !== word[i]) {
        return false;
      }
    }
    return true;
  }
}

export default WordSearch;

File: src/2025-02-24-deepclaude-rust/go/exercises/practice/word-search/word_search.go
package wordsearch

func Solve(words []string, puzzle []string) (map[string][2][2]int, error) {
	result := make(map[string][2][2]int)
	
	// Convert puzzle into a 2D array for easier access
	grid := make([][]rune, len(puzzle))
	for i, row := range puzzle {
		grid[i] = []rune(row)
	}
	
	// For each word, try to find it in the grid
	for _, word := range words {
		// Try each starting position
		for row := 0; row < len(grid); row++ {
			for col := 0; col < len(grid[0]); col++ {
				// Try each direction
				for _, dRow := range []int{-1, 0, 1} {
					for _, dCol := range []int{-1, 0, 1} {
						// Skip (0,0) direction as it's not a valid search direction
						if dRow == 0 && dCol == 0 {
							continue
						}
						
						// Check if word can fit in this direction
						endRow := row + dRow*(len(word)-1)
						endCol := col + dCol*(len(word)-1)
						
						// Check bounds
						if endRow >= 0 && endRow < len(grid) &&
							endCol >= 0 && endCol < len(grid[0]) {
							
							// Try to match word in this direction
							match := true
							for i := 0; i < len(word); i++ {
								checkRow := row + dRow*i
								checkCol := col + dCol*i
								if rune(word[i]) != grid[checkRow][checkCol] {
									match = false
									break
								}
							}
							
							if match {
								// Word found! Store coordinates of first and last letter
								result[word] = [2][2]int{{row, col}, {endRow, endCol}}
							}
						}
					}
				}
			}
		}
	}
	
	return result, nil
}
